geoutils/raster/delayed.py

```diff
@@ -0,0 +1,777 @@
+"""
+Module for dask-delayed functions for out-of-memory raster operations.
+"""
+from __future__ import annotations
+
+import warnings
+from typing import Any, Literal, TypeVar
+
+import dask.array as da
+import dask.delayed
+import geopandas as gpd
+import numpy as np
+import pandas as pd
+import rasterio as rio
+from scipy.interpolate import interpn
+
+from geoutils._typing import NDArrayNum
+from geoutils.projtools import _get_bounds_projected, _get_footprint_projected
+
+# 1/ SUBSAMPLING
+# Getting an exact subsample size out-of-memory only for valid values is not supported directly by Dask/Xarray
+
+# It is not trivial because we don't know where valid values will be in advance, and because of ragged output (varying
+# output length considerations), which prevents from using high-level functions with good efficiency
+# We thus follow https://blog.dask.org/2021/07/02/ragged-output (the dask.array.map_blocks solution has a larger RAM
+# usage by having to drop an axis and re-chunk along 1D of the 2D array, so we use the dask.delayed solution instead)
+
+
+def _random_state_from_user_input(
+    random_state: np.random.RandomState | int | None = None,
+) -> np.random.RandomState | np.random.Generator:
+    """Define random state based on varied user input."""
+
+    # Define state for random sampling (to fix results during testing)
+    # Not using the legacy random call is crucial for RAM usage: https://github.com/numpy/numpy/issues/14169
+    if random_state is None:
+        rnd: np.random.RandomState | np.random.Generator = np.random.default_rng()
+    elif isinstance(random_state, np.random.RandomState):
+        rnd = random_state
+    else:
+        rnd = np.random.default_rng(seed=42)
+
+    return rnd
+
+
+def _get_subsample_size_from_user_input(subsample: int | float, total_nb_valids: int) -> int:
+    """Get subsample size based on a user input of either integer size or fraction of the number of valid points."""
+
+    # If value is between 0 and 1, use a fraction
+    if (subsample <= 1) & (subsample > 0):
+        npoints = int(subsample * total_nb_valids)
+    # Otherwise use the value directly
+    elif subsample > 1:
+        # Use the number of valid points if larger than subsample asked by user
+        npoints = min(int(subsample), total_nb_valids)
+        if subsample > total_nb_valids:
+            warnings.warn(
```

Comment:
Is it really needed? Or maybe we should have an option to silence it as it could show up frequently...

Reply:
I think it's critical to have it especially on an out-of-memory dataset where you can't quickly check how many valid values there are. If nothing is raised, the user would wonder why the length of the subsample returned is different.
Added an option to silence.